### PR TITLE
Change transport connectivity watch API to not be lossy.

### PR DESCRIPTION
**Pull request by markdroth**
This ensures that we don't miss state transitions in LB policies.

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/20255)

### CODE REVIEW

**Comment by AspirinSJL** [comment]
Sorry for delaying. I will review it tomorrow.

File: src/core/ext/filters/max_age/max_age_filter.cc [hunk 1]
```diff
@@ -217,6 +213,46 @@ static void start_max_idle_timer_after_init(void* arg, grpc_error* error) {
...
+  }
+
+  ~ConnectivityWatcher() {
+    GRPC_CHANNEL_STACK_UNREF(chand_->channel_stack, "max_age conn_watch");
+  }
+
+ private:
+  void OnConnectivityStateChange(grpc_connectivity_state new_state) override {
+    if (new_state != GRPC_CHANNEL_SHUTDOWN) return;
+    gpr_mu_lock(&chand_->max_age_timer_mu);
```
Comment (soheilhy) [inline src/core/ext/filters/max_age/max_age_filter.cc]: nit: I would suggest using `MutexLock` here.

Comment (markdroth) [reply src/core/ext/filters/max_age/max_age_filter.cc]: Done.

**Comment by markdroth** [comment]
Known issues: #20383 #20384 #20385

**Pull request merged**